### ISSUE TITLE
[Feature Request] Inverse Hyperbolic Functions

### ISSUE THREAD

**Issue by shawnjhenry** [pytorch/pytorch#10324]
Hi,

As far as I can tell, PyTorch doesn't have inverses implemented for hyperbolic functions.  These are useful, e.g., for operations on hyperbolic space.  (A notable example being that the distance function on the hyperboloid model of hyperbolic space uses the inverse hyperbolic cosine.)  It's not that hard to implement these functions by hand, but inconvenient.

Thanks,
Shawn

cc @VitalyFedyunin @ngimel @mruberry

**Comment by zou3519** [comment]
We have no background on what you're talking about. Could you provide some examples and use cases?

**Comment by shawnjhenry** [comment]
Hi,

See, for example, this paper on learning hierarchical relationships given unstructured similarity scores: https://arxiv.org/pdf/1806.03417.pdf. The authors embed various items (e.g. words in WordNet, persons at Enron, languages) in the hyperboloid model of hyperbolic space to learn hierarchical relations between them in an unsupervised fashion.  The computation of the distance function in the hyperboloid model requires the inverse hyperbolic cosine, which is not currently implemented in PyTorch.  (I would expect it since inverses are implemented for regular trig functions.)  It would be convenient to have these for implementing models like this.  

Another use case would be in models that use normalizing flows (sequences of invertible transformations) as in this paper: https://d4mucfpksywv.cloudfront.net/research-covers/glow/paper/glow.pdf.  When using the model to generate by sampling from the latent space and passing the sample through the inverse network, if one chose to use hyperbolic tangent activations, the inverse network computation would require the inverse hyperbolic tangent.

Regards,
Shawn

**Comment by li012589** [comment]
Hi, I also came across these problems recently. So it is actually easy to code the inverse hyperbolic functions according to Wikipedia (https://en.wikipedia.org/wiki/Inverse_hyperbolic_functions). 

e.g.
```python
def asinh(x):
    return torch.log(x+(x**2+1)**0.5)

def acosh(x):
    return torch.log(x+(x**2-1)**0.5)

def atanh(x):
    return 0.5*torch.log((1+x)/(1-x))
```

Though, I think pytorch should include these functions.

**Comment by shawnjhenry** [comment]
The issue with implementation is that the straightforward version as in the previous comment is not numerically stable.  It would be good to have a numerically stable implementation available.

**Comment by zuoxingdong** [comment]
@zou3519 @vishwakftw Any update on this?

**Comment by vishwakftw** [comment]
As far as I am aware, no one is taking care of this at the moment.

**Comment by zuoxingdong** [comment]
@vishwakftw Thanks for the quick response. Do you know which file in the repo could be the place for a PR to support these functions?

**Comment by cpuhrsch** [comment]
Unary operations like this are implemented in [UnaryOps.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/UnaryOps.cpp#L310). But be warned that it's not necessarily easy to implement this, since this lives relatively far within in the core and even more so if you care about numerical stability.

**Comment by VitalyFedyunin** [comment]
We will gladly accept PR which implements this functions (with numerical stability in mind). But we have no plans to work ourself.

**Comment by aribenjamin** [comment]
A more numerically stable atanh one-liner is `torch.log1p(2*x/(1-x)) / 2` 

From https://www.plunk.org/~hatch/rightway.php

**Comment by SsnL** [comment]
Implemented by https://github.com/pytorch/pytorch/pull/38388
